### PR TITLE
Cleanup IOleObject interface

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Cleanup `IOleObject` interface

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2356)

### CODE REVIEW

File: src/Common/src/UnsafeNativeMethods.cs [hunk 1]
```diff
@@ -877,22 +874,19 @@ public unsafe interface IOleObject
...
 
             [PreserveSig]
-            int GetUserType(
-                   [In, MarshalAs(UnmanagedType.U4)]
-                     int dwFormOfType,
-                   [Out, MarshalAs(UnmanagedType.LPWStr)]
-                     out string userType);
+            HRESULT GetUserType(
+                Ole32.USERCLASSTYPE dwFormOfType,
+                out string userType);
```
Comment (weltkante) [inline src/Common/src/UnsafeNativeMethods.cs]: This is an `LPOLESTR` allocated by the called function, doesn't `string` default to `BSTR`? I don't think you can remove the `MarshalAs` for the string?

Also the [doc](https://docs.microsoft.com/en-us/windows/win32/api/oleidl/nf-oleidl-ioleobject-getusertype) says

> The caller must free pszUserType using the current IMalloc instance.

Will .NET Core do the right thing when marking the string as `MarshalAs(LPWStr)` ?

/cc @JeremyKuhne

File: src/Common/src/UnsafeNativeMethods.cs [hunk 2]
```diff
@@ -990,22 +981,19 @@ public unsafe interface IOleInPlaceObjectWindowless
...
 
             [PreserveSig]
-            int GetUserType(
-                   [In, MarshalAs(UnmanagedType.U4)]
-                     int dwFormOfType,
-                   [Out, MarshalAs(UnmanagedType.LPWStr)]
-                     out string userType);
+            HRESULT GetUserType(
+                Ole32.USERCLASSTYPE dwFormOfType,
+                out string pszUserType);
```
Comment (weltkante) [inline src/Common/src/UnsafeNativeMethods.cs]: ditto

File: src/Common/src/UnsafeNativeMethods.cs (hunk 1 above)
Comment (hughbe) [reply src/Common/src/UnsafeNativeMethods.cs]: Not sure about the allocation part, but I've marked it as `MarshalAs(LPWStr)` now :)

**Comment by RussKie** [comment]
@weltkante do you have further objections?

**Comment by weltkante** [comment]
> > The caller must free pszUserType using the current IMalloc instance.
>
> Will .NET Core do the right thing when marking the string as MarshalAs(LPWStr) ?

I think .NET will do the right thing, at least as long as nobody installed a custom IMalloc, I think both .NET Core and the default IMalloc will use FreeCoTaskMem - but I don't know where to look in the docs to verify it. Also I have no idea how compatible this would be with a custom IMalloc installed.

So not a direct objection since the old code was using LPWStr as well, it would just have been nice to confirm with @JeremyKuhne but if he's not available feel free to merge, it won't introduce any new bugs which weren't already there.

**Comment by RussKie** [comment]
@JeremyKuhne should be back next week. I'll leave it open until then.

**Comment by JeremyKuhne** [comment]
I'm having a heck of a time finding the docs for how marshaling frees strings.  We do have this: https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/interop-pinvokes.md#windows-specific

@jkoritzinsky Do you know where we document string cleanup? We should probably try to make it easier to find. :)

I believe LPWSTR _always_ goes through the Unicode string [marshal code](https://github.com/dotnet/runtime/blob/0b17de659022f8478cca92487f3eaf5dddcfad22/src/coreclr/src/tools/Common/TypeSystem/Interop/IL/Marshaller.cs#L1409). You can see the call to `Marshal.FreeCoTaskMem` there and you can confirm by dumping the interop IL or setting a breakpoint on `Marshal.FreeCoTaskMem`.

**Comment by weltkante** [comment]
Thanks for the [reference](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/interop-pinvokes.md#windows-specific)

Just for completeness if anyone comes by and wonders

> Also I have no idea how compatible this would be with a custom `IMalloc` installed.

`IMalloc` and `CoTaskMemAlloc`/`CoTaskMemFree` are equivalent, see [doc](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-cotaskmemalloc) where it says:

> CoTaskMemAlloc uses the default allocator to allocate a memory block in the same way that IMalloc::Alloc does. It is not necessary to call the CoGetMalloc function before calling CoTaskMemAlloc.

So obviously the same must be true for `CoTaskMemFree`.

**Pull request merged**